tower-retry/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+#[macro_use]
+extern crate futures;
+extern crate tower_service;
+
+use futures::{Async, Future, Poll};
+use tower_service::Service;
+
+#[derive(Clone, Debug)]
+pub struct Retry<P, S> {
+    policy: P,
+    service: S,
+}
+
+#[derive(Debug)]
+pub struct ResponseFuture<P: Policy<S::Request, S::Response, S::Error>, S: Service> {
+    request: Option<S::Request>,
+    retry: Retry<P, S>,
+    state: State<S::Future, P::Future, S::Response, S::Error>,
+}
+
+#[derive(Debug)]
+enum State<F, P, R, E> {
+    /// Polling the future from `Service::call`
+    Called(F),
+    /// Polling the future from `Policy::retry`
+    Checking(P, Option<Result<R, E>>),
+    /// Polling `Service::poll_ready` after `Checking` was OK.
+    Retrying,
+}
+
+pub trait Policy<Req, Res, E>: Sized {
+    type Future: Future<Item=Self, Error=()>;
+    fn retry(&self, req: &Req, res: Result<&Res, &E>) -> Option<Self::Future>;
+    fn clone_request(&self, req: &Req) -> Option<Req>;
```

Comment:
I would recommend a different name than `clone_request`. The `clone` name implies that the return value is identical to the source. Instead, the fn might want to modify the request somehow. For example, it might add an HTTP header w/ the attempt count.

Reply:
Any name recommendations?